### PR TITLE
Feature/callbacks optional keys

### PR DESCRIPTION
**Pull request by benjamin-work**
Main changes:
* Callbacks now have an `keys_optional` keyword that allows to exclude keys from check.
* An additional check for duplicate keys has been added.
* Error messages have been improved.

*Note*: The idea of telling the user exactly how to `set_params` to a add a key to `keys_optional` does not work. For that, we would need to know the name of the callback, since the name is part of the `set_params` path. However, we don't know the callback name at the time of the call.

### CODE REVIEW

File: inferno/callbacks.py [hunk 1]
```diff
@@ -172,12 +217,30 @@ class BestLoss(Callback):
...
-    def initialize(self):
+        self._check_keys_duplicated()
+        self._check_signs()
+
+    def _is_optional(self, key):
+        return (key in self.keys_optional) or (key in self.default_key_signs)
+
+    def _check_keys_duplicated(self):
+        duplicates = duplicate_items(self.default_key_signs, self.key_signs)
+        if duplicates:
```
Comment (ottonemo) [inline inferno/callbacks.py]: this can be merged into one line

File: inferno/callbacks.py [hunk 2]
```diff
@@ -109,35 +111,70 @@ class AverageLoss(Callback):
...
+                    yield key_loss, list(zip(*row))
+                except KeyError:
+                    pass
+                continue
             row = history[-1, 'batches', :, (key_loss, key_size)]
             yield key_loss, list(zip(*row))
 
+    def _check_keys_duplicated(self):
+        duplicates = duplicate_items(self.default_key_sizes, self.key_sizes)
+        if duplicates:
```
Comment (ottonemo) [inline inferno/callbacks.py]: this can be merged into one line

File: inferno/callbacks.py [hunk 3]
```diff
@@ -109,35 +111,70 @@ class AverageLoss(Callback):
...
+        for key_loss, key_size in key_sizes:
+            if self._is_optional(key_loss) or self._is_optional(key_size):
+                try:
+                    row = history[-1, 'batches', :, (key_loss, key_size)]
+                    yield key_loss, list(zip(*row))
+                except KeyError:
+                    pass
+                continue
             row = history[-1, 'batches', :, (key_loss, key_size)]
             yield key_loss, list(zip(*row))
```
Comment (ottonemo) [inline inferno/callbacks.py]: This whole inner-loop block can be simplified to

```
try:
    row = ...
    yield ...
except KeyError:
    if not self._is_optional(key_loss) and not self._is_optional(key_size):
        raise
```

File: inferno/callbacks.py [hunk 4]
```diff
@@ -186,24 +249,32 @@ def initialize(self):
...
+        for key_loss, sign in key_signs:
+            if self._is_optional(key_loss):
+                try:
+                    loss = history[-1, key_loss]
+                    yield key_loss, sign, loss
+                except KeyError:
+                    pass
+                continue
+            loss = history[-1, key_loss]
+            yield key_loss, sign, loss
```
Comment (ottonemo) [inline inferno/callbacks.py]: this can be simplified similar to the block above (i.e. raise on error, try always)

File: inferno/callbacks.py [hunk 5]
```diff
@@ -303,6 +374,14 @@ class PrintLog(Callback):
       on `'_best'` are used to determine the best corresponding loss
       (see above).
 
+    keys_optional: list of str or None (default=None)
+      If not None, this should be a list of keys whose presence is
+      optional. By default, keys indicated in `key` are mandatory, but
```
Comment (ottonemo) [inline inferno/callbacks.py]: "are mandatory [to exist in the history]"

File: inferno/callbacks.py [hunk 6]
```diff
@@ -377,8 +481,16 @@ def table(self, row):
...
         for key, color in zip(self._yield_keys(row), colors):
+            if self._is_optional(key):
+                try:
+                    headers.append(key)
+                    formatted.append(self.format_row(row, key, color=color))
+                except KeyError:
+                    pass
+                continue
             headers.append(key)
             formatted.append(self.format_row(row, key, color=color))
```
Comment (ottonemo) [inline inferno/callbacks.py]: This can be simplified similar to the blocks above

File: inferno/callbacks.py [hunk 7]
```diff
@@ -341,14 +421,38 @@ def __init__(
...
         self.floatfmt = floatfmt
 
+        self._check_keys_duplicated()
+
+    def _is_optional(self, key):
+        return (key in self.keys_optional) or (key in self.default_keys)
+
+    def _check_keys_duplicated(self):
+        duplicates = duplicate_items(self.default_keys, self.keys)
+        if duplicates:
```
Comment (ottonemo) [inline inferno/callbacks.py]: this can be merged into one line

File: inferno/callbacks.py [hunk 8]
```diff
@@ -109,35 +111,70 @@ class AverageLoss(Callback):
...
 
+    def _check_keys_duplicated(self):
+        duplicates = duplicate_items(self.default_key_sizes, self.key_sizes)
+        if duplicates:
+            raise ValueError("AverageLoss found duplicate keys: {}"
+                             "".format(', '.join(sorted(duplicates))))
+
+    def _check_keys_missing(self, history):
+        check_keys = []
+        for key in flatten(self.key_sizes.items()):
```
Comment (ottonemo) [inline inferno/callbacks.py]: Maybe add an explanation why we use `flatten(key_sizes.items())`, e.g.

    # Check value keys and size keys alike.

File: inferno/callbacks.py (hunk 2 above)
Comment (benjamin-work) [inline inferno/callbacks.py]: duplicate assignment is needed later, thus won't fix.

File: inferno/callbacks.py (hunk 3 above)
Comment (benjamin-work) [inline inferno/callbacks.py]: much better, ty :+1:

File: inferno/callbacks.py (hunk 1 above)
Comment (benjamin-work) [inline inferno/callbacks.py]: won't fix (s. above)

File: inferno/callbacks.py (hunk 7 above)
Comment (benjamin-work) [inline inferno/callbacks.py]: won't fix (s. above)

**Comment by benjamin-work** [comment]
@ottonemo please look again

**Pull request merged**